rskj-core/src/main/java/co/rsk/core/bc/TransactionConflictDetector.java

```diff
@@ -0,0 +1,367 @@
+            // because in such a case, the readers are already in conflict with it.
+            checkKeyAlreadyRead(key, partitionId);
+        }
+        checkAccountDeleted(cacheTracking.getAccountFromKey(key), partitionId);
+        // Whatever if there is a conflict or not, track access to the account of this key
+        trackAccessToAccount(cacheTracking.getAccountFromKey(key), partitionId);
+    }
+
+    @Override
+    public synchronized void onDeleteAccount(ICacheTracking cacheTracking, ByteArrayWrapper account, int partitionId) {
+        // There is a conflict if :
+        // - some keys of this account have been accessed (read or written) by another partition
+        // or
+        // - the account have already been deleted by another partition
+
+        checkAccountAccessed(account, partitionId);
+        if (!checkAccountDeleted(account, partitionId)) {
+            trackAccountDeleted(account, partitionId);
+        }
+    }
+
+    private Conflict createConflict(ByteArrayWrapper key, int conflictPartId, int originPartId, eConflictType conflictType) {
+        TransactionsPartition conflictPartition = partitioner.fromId(conflictPartId);
+        TransactionsPartition originPartition = partitioner.fromId(originPartId);
+        if (conflictPartition == null) {
+            logger.error("Unable to get the partition for id [{}]", conflictPartId);
+        }
+        if (originPartition == null) {
+            logger.error("Unable to get the partition for id [{}]", originPartId);
+        }
+        return new Conflict(key, conflictPartition, originPartition, conflictType);
+    }
+
+    protected void recordConflict(Conflict conflict) {
+        Collection<Conflict> conflicts;
+        conflicts = conflictsPerPartition.computeIfAbsent(conflict.getConflictFrom(), k -> new HashSet<>());
+        conflicts.add(conflict);
+        conflicts = conflictsPerPartition.computeIfAbsent(conflict.getConflictWith(), k -> new HashSet<>());
+        conflicts.add(conflict);
+    }
+
+    private boolean checkKeyAlreadyWritten(ByteArrayWrapper key, int partId) {
+        Integer otherPartId = partitionWriterPerKey.get(key);
+        if ((otherPartId == null) || otherPartId.equals(partId)) {
+            return false;
+        } else {
+            logger.info("key [{}] has already been written by partition [{}] --> conflict from partition [{}]",
+                    key, otherPartId, partId);
+            recordConflict(createConflict(key, otherPartId, partId, eConflictType.WRITTEN_BEFORE_ACCESSED));
+        }
+        return true;
+    }
+
+    private boolean checkKeyAlreadyRead(ByteArrayWrapper key, int partitionId) {
+        Collection<Integer> otherPartIds = partitionReadersPerKey.get(key);
+        if (otherPartIds == null) {
+            return false;
+        }
+        // there can not be a conflict with itself
+        otherPartIds.remove(partitionId);
```

Comment:
This remove() actually modifies the underlying data structure so has side effects. It should either re-insert the partitionId or don't remove it in the first place and check for size()==1 and the first element.

Reply:
@SergioDemianLerner  Well done !
You’re right, there is a bug here.
Following your comment, I’ve tried to find a usecase which reveals it, and add a test case and check the correction fix it effectively.
But I couldn’t find this case: this bug has actually no real impact because the only case when you remove the current partition from the readers list of the key is when the same partition is writing this key too.
So, in the end there will be a conflict with any other partition that would attempt to access the same key, even if the readers list has been purged (unintentionally I confess).
Anyway, I make the fix. This should have a positive impact on performances, however.